### DISCUSSION TITLE
Wrapping pytest call with default test path

### DISCUSSION THREAD

**Discussion by Novakov** [pytest-dev/pytest#8801] [Q&A]
Project I'm working on is using wrapper around pytest that automatically generates few required custom parameters and we recently found issue with explicit file paths. As I was not able to narrow down the problem to simple & small use case, repository https://github.com/Novakov/pytest-path-problem contains detailed description and reproduction of the issue.

It can be summarised with: how can I tell pytest to use directory X if no other directory/file is provided (so call `pytest --base-dir=X` is equal to `pytest X` and `pytest --base-dir=X file1.py` is equal to `pytest file1.py`).

It is quite possible that it is not possible to do (and I'm okay with that) but I'm not giving up hope yet, any suggestion what I can try is welcome.

**Comment by portyu9** [comment]
Hi @Novakov 

The wrapper should not always append `hw_tests` as a positional argument.

pytest already has the behavior you want through `testpaths`: it supplies default test paths when no explicit test target is given, while an explicit file/directory on the command line takes precedence.

Put the default in pytest configuration, for example:

```ini
[pytest]
testpaths = hw_tests
```

and make the wrapper invoke pytest **without** adding `hw_tests` itself:

```python
cmdline = [
    sys.executable,
    "-m",
    "pytest",
    "--argument",
    "abc",
    *sys.argv[1:],
]
```

Because your wrapper should work from arbitrary directories, also run pytest with the project/config directory as its working directory:

```python
subprocess.run(cmdline, cwd=project_root)
```

Then:

```text
do_test
```

uses `testpaths`, while:

```text
do_test hw_tests/test_hw_1.py
```

uses only that explicitly supplied test file, so it is not collected once through the directory and a second time through the file.

This also gives pytest a consistent place from which to discover the root configuration/conftest containing `--argument`.

Please mark this answer as accepted if it helped, thank you.
